pkg/skaffold/build/scheduler.go

```diff
@@ -35,45 +35,58 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
 
 // For testing
 var (
-	buffSize      = bufferedLinesPerArtifact
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
+type artifactWithDeps struct {
+	artifact              *latest.Artifact
+	artifactChan          chan interface{}
+	requiredArtifactChans []chan interface{}
+}
 
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
+func (a artifactWithDeps) markComplete() {
+	// closing channel notifies all listeners waiting for this build to complete
+	close(a.artifactChan)
+}
+func (a artifactWithDeps) waitForDependencies(ctx context.Context) {
+	for _, dep := range a.requiredArtifactChans {
+		// wait for dependency to complete build
+		select {
+		case <-ctx.Done():
+		case <-dep:
+		}
 	}
+}
 
+// InOrder builds a list of artifacts in dependency order within the specified max concurrency
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
+	awdSlice := artifactWithDepsFor(artifacts)
 	var wg sync.WaitGroup
 	defer wg.Wait()
 
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	results := new(sync.Map)
-	outputs := make([]chan string, len(artifacts))
+	outputs := make([]chan string, len(awdSlice))
 
 	if concurrency == 0 {
-		concurrency = len(artifacts)
+		concurrency = len(awdSlice)
```

Comment:
just trying to work out how this works with nested dependencies. at a certain point, we can't really build concurrently can we? i suppose all top-level dependencies for a given artifact can be built in parallel if (and only if) those don't have dependencies of their own right?

Reply:
yes, I'll add some detail into how I modeled the DAG using a set of channels which allows the max number of parallel builds within `concurrency` for independent artifacts while maintaining required order.

Reply:
added description:
```
//artifactWithDeps models the artifact dependency graph using a set of channels.
// Each artifact has a channel that it closes once it completes building (either success or failure) by calling `markComplete`
// Additionally it has a list of channels for each of its dependencies.
// Calling `waitForDependencies` ensures that all required artifacts' channels have already been closed and as such have finished building.
// This model allows for composing any arbitrary function with dependency ordering.
```
So now additionally piping it through a `sem` buffered channel ensures that at most `sem` number of parallel builds can run if there are no dependencies. But the `waitForDependencies` ensures that all builds need to follow defined order.

PS. go channels are awesome. 🦦

Reply:
Maybe a few more details are required as there are two key numbers at play here: 
  1. _concurrency_ which is the maximum number of builds that should be running at any one time, and
  2. _len(artifacts)_ is the number of artifacts to be built

`awdSlice` is of size _len(artifacts)_.  This code creates a goroutine for each element in `awdSlice`.  Each goroutine waits on its dependencies to finish building.  Because our artifacts form a DAG, at least one of the goroutines should be able to start building.

`markComplete()` closes the artifact's channel **which wakes all of the listeners waiting on that channel**.

Reply:
added all these details in either method description or inline comments.